### DISCUSSION TITLE
Deta Base Delete using filter query.

### DISCUSSION THREAD

**Discussion by harshitsinghai77** [deta/cloud-docs#236] [Help] [answered]
Acc. to docs

```python
delete(key: str)
```

But what if I need to delete all the records having the same `user_id`? Something like, 

![image](https://user-images.githubusercontent.com/30886142/171353688-261d939f-0b8b-4e9e-8d63-e95f33cfa328.png)

```python
deta_task_db.delete( {"user_id": "Harshit" } )
```

Can I do that?

**Comment by BetaHuhn** [comment, marked as answer]
Hi @harshitsinghai77,

Deta currently doesn't offer a `deleteMany` method, so you need to manually delete each item.

What you can do for your example is fetch all records with the same `user_id` and then loop through the returned records and delete each one individually by their key:

```python
records = db.fetch({ "user_id": "Harshit" })

for record in records:
   db.delete(record.key)
```

Hope this helps!

**Reply by harshitsinghai77** [reply]
hmm... I was thinking the same thing. But I think we should have a `deleteMany` method in the future, 

If I fetch records, that's 1 quey, if I have multiple such records, then that's a couple of queries to fetch all records using fetch(last=key), on top of that there will be multiple delete requests (we're not doing batch request, each request is individual). That sounds like a lot of work behind the scenes.

It's slow too (at least in theory).

**Reply by abdelhai** [reply]
we will think about this in our next release. thanks for the feedback!

**Comment by jnsougata** [comment]
However if you have the list of keys to be deleted then you can do delete request asynchronously to the api which will be faster.
Check HTTP API (DELETE): https://docs.deta.sh/docs/base/http#delete-item
USE: asyncio, aiohttp

example:
```python
import aiohttp
import asyncio


async def del_key(session, url):
    async with session.delete(url) as resp:
        return resp.status


async def main():
    async with aiohttp.ClientSession() as session:
        tasks = []
        for key in target_keys:
            url = f'https://database.deta.sh/v1/{project_id}/{base_name}/items/{key}'
            tasks.append(asyncio.create_task(del_key(session, url)))
        done = await asyncio.gather(*tasks)
        for status_code in done:
            print(status_code)

asyncio.run(main())
```
** `assuming DETA doesn't have ratelimits to API calls`

**Reply by abdelhai** [reply]
we don't

**Reply by harshitsinghai77** [reply]
Thanks for the code snippet.

**Reply by harshitsinghai77** [reply]
wieldy, this is still very slow. As slow as sequentially making `delete` request. 

I ran a for loop to create 200 entries in the database, then used aiohttp and asyncio to create multiple delete requests. It took longer than I expected to delete all the entries.

**Reply by harshitsinghai77** [reply]
Also, I'm getting this after a while

![image](https://user-images.githubusercontent.com/30886142/171398550-63546928-af15-4f78-9db9-a09def477086.png)

**Reply by jnsougata** [reply]
```python
   #   This is how I delete many entries asynchronously using HTTP API.
    async def _delete(self, base_name: str, key: str):
        ep = self.__base_root + base_name + '/items/' + key
        await self.__session.delete(ep, headers=self.__base_headers)
        return key

    async def _delete_many(self, base_name: str, keys: list):
        task = [asyncio.create_task(self._delete(base_name, key)) for key in keys]
        return await asyncio.gather(*task)
  ```

  ```python
 # my driver code
  async def main():
    deta = Deta(os.getenv("DETA_TOKEN"))
    await deta.connect()
    base = deta.base(name='123BASE')
    keys = [f'TARGET{i}' for i in range(1000)]
    s = time.perf_counter()
    print(await base.delete_many(keys))
    e = time.perf_counter()
    await deta.close()
    print(f'Time taken {e - s:.2f} seconds to delete {len(keys)} items')


loop = asyncio.new_event_loop()
loop.run_until_complete(main())
```
`Result: Takes almost 3 seconds to delete 1000 entries!`

`/home/jnsougata/Git/async-deta/venv/bin/python /home/jnsougata/Git/async-deta/testing/main.py`
['TARGET0', 'TARGET1', 'TARGET2', 'TARGET3', 'TARGET4', 'TARGET5', 'TARGET6', 'TARGET7', 'TARGET8', 'TARGET9', 'TARGET10', 'TARGET11', 'TARGET12', 'TARGET13', 'TARGET14', 'TARGET15', 'TARGET16', 'TARGET17', 'TARGET18', 'TARGET19', 'TARGET20', 'TARGET21', 'TARGET22', 'TARGET23', 'TARGET24', 'TARGET25', 'TARGET26', 'TARGET27', 'TARGET28', 'TARGET29', 'TARGET30', 'TARGET31', 'TARGET32', 'TARGET33', 'TARGET34', 'TARGET35', 'TARGET36', 'TARGET37', 'TARGET38', 'TARGET39', 'TARGET40', 'TARGET41', 'TARGET42', 'TARGET43', 'TARGET44', 'TARGET45', 'TARGET46', 'TARGET47', 'TARGET48', 'TARGET49', 'TARGET50', 'TARGET51', 'TARGET52', 'TARGET53', 'TARGET54', 'TARGET55', 'TARGET56', 'TARGET57', 'TARGET58', 'TARGET59', 'TARGET60', 'TARGET61', 'TARGET62', 'TARGET63', 'TARGET64', 'TARGET65', 'TARGET66', 'TARGET67', 'TARGET68', 'TARGET69', 'TARGET70', 'TARGET71', 'TARGET72', 'TARGET73', 'TARGET74', 'TARGET75', 'TARGET76', 'TARGET77', 'TARGET78', 'TARGET79', 'TARGET80', 'TARGET81', 'TARGET82', 'TARGET83', 'TARGET84', 'TARGET85', 'TARGET86', 'TARGET87', 'TARGET88', 'TARGET89', 'TARGET90', 'TARGET91', 'TARGET92', 'TARGET93', 'TARGET94', 'TARGET95', 'TARGET96', 'TARGET97', 'TARGET98', 'TARGET99', 'TARGET100', 'TARGET101', 'TARGET102', 'TARGET103', 'TARGET104', 'TARGET105', 'TARGET106', 'TARGET107', 'TARGET108', 'TARGET109', 'TARGET110', 'TARGET111', 'TARGET112', 'TARGET113', 'TARGET114', 'TARGET115', 'TARGET116', 'TARGET117', 'TARGET118', 'TARGET119', 'TARGET120', 'TARGET121', 'TARGET122', 'TARGET123', 'TARGET124', 'TARGET125', 'TARGET126', 'TARGET127', 'TARGET128', 'TARGET129', 'TARGET130', 'TARGET131', 'TARGET132', 'TARGET133', 'TARGET134', 'TARGET135', 'TARGET136', 'TARGET137', 'TARGET138', 'TARGET139', 'TARGET140', 'TARGET141', 'TARGET142', 'TARGET143', 'TARGET144', 'TARGET145', 'TARGET146', 'TARGET147', 'TARGET148', 'TARGET149', 'TARGET150', 'TARGET151', 'TARGET152', 'TARGET153', 'TARGET154', 'TARGET155', 'TARGET156', 'TARGET157', 'TARGET158', 'TARGET159', 'TARGET160', 'TARGET161', 'TARGET162', 'TARGET163', 'TARGET164', 'TARGET165', 'TARGET166', 'TARGET167', 'TARGET168', 'TARGET169', 'TARGET170', 'TARGET171', 'TARGET172', 'TARGET173', 'TARGET174', 'TARGET175', 'TARGET176', 'TARGET177', 'TARGET178', 'TARGET179', 'TARGET180', 'TARGET181', 'TARGET182', 'TARGET183', 'TARGET184', 'TARGET185', 'TARGET186', 'TARGET187', 'TARGET188', 'TARGET189', 'TARGET190', 'TARGET191', 'TARGET192', 'TARGET193', 'TARGET194', 'TARGET195', 'TARGET196', 'TARGET197', 'TARGET198', 'TARGET199', 'TARGET200', 'TARGET201', 'TARGET202', 'TARGET203', 'TARGET204', 'TARGET205', 'TARGET206', 'TARGET207', 'TARGET208', 'TARGET209', 'TARGET210', 'TARGET211', 'TARGET212', 'TARGET213', 'TARGET214', 'TARGET215', 'TARGET216', 'TARGET217', 'TARGET218', 'TARGET219', 'TARGET220', 'TARGET221', 'TARGET222', 'TARGET223', 'TARGET224', 'TARGET225', 'TARGET226', 'TARGET227', 'TARGET228', 'TARGET229', 'TARGET230', 'TARGET231', 'TARGET232', 'TARGET233', 'TARGET234', 'TARGET235', 'TARGET236', 'TARGET237', 'TARGET238', 'TARGET239', 'TARGET240', 'TARGET241', 'TARGET242', 'TARGET243', 'TARGET244', 'TARGET245', 'TARGET246', 'TARGET247', 'TARGET248', 'TARGET249', 'TARGET250', 'TARGET251', 'TARGET252', 'TARGET253', 'TARGET254', 'TARGET255', 'TARGET256', 'TARGET257', 'TARGET258', 'TARGET259', 'TARGET260', 'TARGET261', 'TARGET262', 'TARGET263', 'TARGET264', 'TARGET265', 'TARGET266', 'TARGET267', 'TARGET268', 'TARGET269', 'TARGET270', 'TARGET271', 'TARGET272', 'TARGET273', 'TARGET274', 'TARGET275', 'TARGET276', 'TARGET277', 'TARGET278', 'TARGET279', 'TARGET280', 'TARGET281', 'TARGET282', 'TARGET283', 'TARGET284', 'TARGET285', 'TARGET286', 'TARGET287', 'TARGET288', 'TARGET289', 'TARGET290', 'TARGET291', 'TARGET292', 'TARGET293', 'TARGET294', 'TARGET295', 'TARGET296', 'TARGET297', 'TARGET298', 'TARGET299', 'TARGET300', 'TARGET301', 'TARGET302', 'TARGET303', 'TARGET304', 'TARGET305', 'TARGET306', 'TARGET307', 'TARGET308', 'TARGET309', 'TARGET310', 'TARGET311', 'TARGET312', 'TARGET313', 'TARGET314', 'TARGET315', 'TARGET316', 'TARGET317', 'TARGET318', 'TARGET319', 'TARGET320', 'TARGET321', 'TARGET322', 'TARGET323', 'TARGET324', 'TARGET325', 'TARGET326', 'TARGET327', 'TARGET328', 'TARGET329', 'TARGET330', 'TARGET331', 'TARGET332', 'TARGET333', 'TARGET334', 'TARGET335', 'TARGET336', 'TARGET337', 'TARGET338', 'TARGET339', 'TARGET340', 'TARGET341', 'TARGET342', 'TARGET343', 'TARGET344', 'TARGET345', 'TARGET346', 'TARGET347', 'TARGET348', 'TARGET349', 'TARGET350', 'TARGET351', 'TARGET352', 'TARGET353', 'TARGET354', 'TARGET355', 'TARGET356', 'TARGET357', 'TARGET358', 'TARGET359', 'TARGET360', 'TARGET361', 'TARGET362', 'TARGET363', 'TARGET364', 'TARGET365', 'TARGET366', 'TARGET367', 'TARGET368', 'TARGET369', 'TARGET370', 'TARGET371', 'TARGET372', 'TARGET373', 'TARGET374', 'TARGET375', 'TARGET376', 'TARGET377', 'TARGET378', 'TARGET379', 'TARGET380', 'TARGET381', 'TARGET382', 'TARGET383', 'TARGET384', 'TARGET385', 'TARGET386', 'TARGET387', 'TARGET388', 'TARGET389', 'TARGET390', 'TARGET391', 'TARGET392', 'TARGET393', 'TARGET394', 'TARGET395', 'TARGET396', 'TARGET397', 'TARGET398', 'TARGET399', 'TARGET400', 'TARGET401', 'TARGET402', 'TARGET403', 'TARGET404', 'TARGET405', 'TARGET406', 'TARGET407', 'TARGET408', 'TARGET409', 'TARGET410', 'TARGET411', 'TARGET412', 'TARGET413', 'TARGET414', 'TARGET415', 'TARGET416', 'TARGET417', 'TARGET418', 'TARGET419', 'TARGET420', 'TARGET421', 'TARGET422', 'TARGET423', 'TARGET424', 'TARGET425', 'TARGET426', 'TARGET427', 'TARGET428', 'TARGET429', 'TARGET430', 'TARGET431', 'TARGET432', 'TARGET433', 'TARGET434', 'TARGET435', 'TARGET436', 'TARGET437', 'TARGET438', 'TARGET439', 'TARGET440', 'TARGET441', 'TARGET442', 'TARGET443', 'TARGET444', 'TARGET445', 'TARGET446', 'TARGET447', 'TARGET448', 'TARGET449', 'TARGET450', 'TARGET451', 'TARGET452', 'TARGET453', 'TARGET454', 'TARGET455', 'TARGET456', 'TARGET457', 'TARGET458', 'TARGET459', 'TARGET460', 'TARGET461', 'TARGET462', 'TARGET463', 'TARGET464', 'TARGET465', 'TARGET466', 'TARGET467', 'TARGET468', 'TARGET469', 'TARGET470', 'TARGET471', 'TARGET472', 'TARGET473', 'TARGET474', 'TARGET475', 'TARGET476', 'TARGET477', 'TARGET478', 'TARGET479', 'TARGET480', 'TARGET481', 'TARGET482', 'TARGET483', 'TARGET484', 'TARGET485', 'TARGET486', 'TARGET487', 'TARGET488', 'TARGET489', 'TARGET490', 'TARGET491', 'TARGET492', 'TARGET493', 'TARGET494', 'TARGET495', 'TARGET496', 'TARGET497', 'TARGET498', 'TARGET499', 'TARGET500', 'TARGET501', 'TARGET502', 'TARGET503', 'TARGET504', 'TARGET505', 'TARGET506', 'TARGET507', 'TARGET508', 'TARGET509', 'TARGET510', 'TARGET511', 'TARGET512', 'TARGET513', 'TARGET514', 'TARGET515', 'TARGET516', 'TARGET517', 'TARGET518', 'TARGET519', 'TARGET520', 'TARGET521', 'TARGET522', 'TARGET523', 'TARGET524', 'TARGET525', 'TARGET526', 'TARGET527', 'TARGET528', 'TARGET529', 'TARGET530', 'TARGET531', 'TARGET532', 'TARGET533', 'TARGET534', 'TARGET535', 'TARGET536', 'TARGET537', 'TARGET538', 'TARGET539', 'TARGET540', 'TARGET541', 'TARGET542', 'TARGET543', 'TARGET544', 'TARGET545', 'TARGET546', 'TARGET547', 'TARGET548', 'TARGET549', 'TARGET550', 'TARGET551', 'TARGET552', 'TARGET553', 'TARGET554', 'TARGET555', 'TARGET556', 'TARGET557', 'TARGET558', 'TARGET559', 'TARGET560', 'TARGET561', 'TARGET562', 'TARGET563', 'TARGET564', 'TARGET565', 'TARGET566', 'TARGET567', 'TARGET568', 'TARGET569', 'TARGET570', 'TARGET571', 'TARGET572', 'TARGET573', 'TARGET574', 'TARGET575', 'TARGET576', 'TARGET577', 'TARGET578', 'TARGET579', 'TARGET580', 'TARGET581', 'TARGET582', 'TARGET583', 'TARGET584', 'TARGET585', 'TARGET586', 'TARGET587', 'TARGET588', 'TARGET589', 'TARGET590', 'TARGET591', 'TARGET592', 'TARGET593', 'TARGET594', 'TARGET595', 'TARGET596', 'TARGET597', 'TARGET598', 'TARGET599', 'TARGET600', 'TARGET601', 'TARGET602', 'TARGET603', 'TARGET604', 'TARGET605', 'TARGET606', 'TARGET607', 'TARGET608', 'TARGET609', 'TARGET610', 'TARGET611', 'TARGET612', 'TARGET613', 'TARGET614', 'TARGET615', 'TARGET616', 'TARGET617', 'TARGET618', 'TARGET619', 'TARGET620', 'TARGET621', 'TARGET622', 'TARGET623', 'TARGET624', 'TARGET625', 'TARGET626', 'TARGET627', 'TARGET628', 'TARGET629', 'TARGET630', 'TARGET631', 'TARGET632', 'TARGET633', 'TARGET634', 'TARGET635', 'TARGET636', 'TARGET637', 'TARGET638', 'TARGET639', 'TARGET640', 'TARGET641', 'TARGET642', 'TARGET643', 'TARGET644', 'TARGET645', 'TARGET646', 'TARGET647', 'TARGET648', 'TARGET649', 'TARGET650', 'TARGET651', 'TARGET652', 'TARGET653', 'TARGET654', 'TARGET655', 'TARGET656', 'TARGET657', 'TARGET658', 'TARGET659', 'TARGET660', 'TARGET661', 'TARGET662', 'TARGET663', 'TARGET664', 'TARGET665', 'TARGET666', 'TARGET667', 'TARGET668', 'TARGET669', 'TARGET670', 'TARGET671', 'TARGET672', 'TARGET673', 'TARGET674', 'TARGET675', 'TARGET676', 'TARGET677', 'TARGET678', 'TARGET679', 'TARGET680', 'TARGET681', 'TARGET682', 'TARGET683', 'TARGET684', 'TARGET685', 'TARGET686', 'TARGET687', 'TARGET688', 'TARGET689', 'TARGET690', 'TARGET691', 'TARGET692', 'TARGET693', 'TARGET694', 'TARGET695', 'TARGET696', 'TARGET697', 'TARGET698', 'TARGET699', 'TARGET700', 'TARGET701', 'TARGET702', 'TARGET703', 'TARGET704', 'TARGET705', 'TARGET706', 'TARGET707', 'TARGET708', 'TARGET709', 'TARGET710', 'TARGET711', 'TARGET712', 'TARGET713', 'TARGET714', 'TARGET715', 'TARGET716', 'TARGET717', 'TARGET718', 'TARGET719', 'TARGET720', 'TARGET721', 'TARGET722', 'TARGET723', 'TARGET724', 'TARGET725', 'TARGET726', 'TARGET727', 'TARGET728', 'TARGET729', 'TARGET730', 'TARGET731', 'TARGET732', 'TARGET733', 'TARGET734', 'TARGET735', 'TARGET736', 'TARGET737', 'TARGET738', 'TARGET739', 'TARGET740', 'TARGET741', 'TARGET742', 'TARGET743', 'TARGET744', 'TARGET745', 'TARGET746', 'TARGET747', 'TARGET748', 'TARGET749', 'TARGET750', 'TARGET751', 'TARGET752', 'TARGET753', 'TARGET754', 'TARGET755', 'TARGET756', 'TARGET757', 'TARGET758', 'TARGET759', 'TARGET760', 'TARGET761', 'TARGET762', 'TARGET763', 'TARGET764', 'TARGET765', 'TARGET766', 'TARGET767', 'TARGET768', 'TARGET769', 'TARGET770', 'TARGET771', 'TARGET772', 'TARGET773', 'TARGET774', 'TARGET775', 'TARGET776', 'TARGET777', 'TARGET778', 'TARGET779', 'TARGET780', 'TARGET781', 'TARGET782', 'TARGET783', 'TARGET784', 'TARGET785', 'TARGET786', 'TARGET787', 'TARGET788', 'TARGET789', 'TARGET790', 'TARGET791', 'TARGET792', 'TARGET793', 'TARGET794', 'TARGET795', 'TARGET796', 'TARGET797', 'TARGET798', 'TARGET799', 'TARGET800', 'TARGET801', 'TARGET802', 'TARGET803', 'TARGET804', 'TARGET805', 'TARGET806', 'TARGET807', 'TARGET808', 'TARGET809', 'TARGET810', 'TARGET811', 'TARGET812', 'TARGET813', 'TARGET814', 'TARGET815', 'TARGET816', 'TARGET817', 'TARGET818', 'TARGET819', 'TARGET820', 'TARGET821', 'TARGET822', 'TARGET823', 'TARGET824', 'TARGET825', 'TARGET826', 'TARGET827', 'TARGET828', 'TARGET829', 'TARGET830', 'TARGET831', 'TARGET832', 'TARGET833', 'TARGET834', 'TARGET835', 'TARGET836', 'TARGET837', 'TARGET838', 'TARGET839', 'TARGET840', 'TARGET841', 'TARGET842', 'TARGET843', 'TARGET844', 'TARGET845', 'TARGET846', 'TARGET847', 'TARGET848', 'TARGET849', 'TARGET850', 'TARGET851', 'TARGET852', 'TARGET853', 'TARGET854', 'TARGET855', 'TARGET856', 'TARGET857', 'TARGET858', 'TARGET859', 'TARGET860', 'TARGET861', 'TARGET862', 'TARGET863', 'TARGET864', 'TARGET865', 'TARGET866', 'TARGET867', 'TARGET868', 'TARGET869', 'TARGET870', 'TARGET871', 'TARGET872', 'TARGET873', 'TARGET874', 'TARGET875', 'TARGET876', 'TARGET877', 'TARGET878', 'TARGET879', 'TARGET880', 'TARGET881', 'TARGET882', 'TARGET883', 'TARGET884', 'TARGET885', 'TARGET886', 'TARGET887', 'TARGET888', 'TARGET889', 'TARGET890', 'TARGET891', 'TARGET892', 'TARGET893', 'TARGET894', 'TARGET895', 'TARGET896', 'TARGET897', 'TARGET898', 'TARGET899', 'TARGET900', 'TARGET901', 'TARGET902', 'TARGET903', 'TARGET904', 'TARGET905', 'TARGET906', 'TARGET907', 'TARGET908', 'TARGET909', 'TARGET910', 'TARGET911', 'TARGET912', 'TARGET913', 'TARGET914', 'TARGET915', 'TARGET916', 'TARGET917', 'TARGET918', 'TARGET919', 'TARGET920', 'TARGET921', 'TARGET922', 'TARGET923', 'TARGET924', 'TARGET925', 'TARGET926', 'TARGET927', 'TARGET928', 'TARGET929', 'TARGET930', 'TARGET931', 'TARGET932', 'TARGET933', 'TARGET934', 'TARGET935', 'TARGET936', 'TARGET937', 'TARGET938', 'TARGET939', 'TARGET940', 'TARGET941', 'TARGET942', 'TARGET943', 'TARGET944', 'TARGET945', 'TARGET946', 'TARGET947', 'TARGET948', 'TARGET949', 'TARGET950', 'TARGET951', 'TARGET952', 'TARGET953', 'TARGET954', 'TARGET955', 'TARGET956', 'TARGET957', 'TARGET958', 'TARGET959', 'TARGET960', 'TARGET961', 'TARGET962', 'TARGET963', 'TARGET964', 'TARGET965', 'TARGET966', 'TARGET967', 'TARGET968', 'TARGET969', 'TARGET970', 'TARGET971', 'TARGET972', 'TARGET973', 'TARGET974', 'TARGET975', 'TARGET976', 'TARGET977', 'TARGET978', 'TARGET979', 'TARGET980', 'TARGET981', 'TARGET982', 'TARGET983', 'TARGET984', 'TARGET985', 'TARGET986', 'TARGET987', 'TARGET988', 'TARGET989', 'TARGET990', 'TARGET991', 'TARGET992', 'TARGET993', 'TARGET994', 'TARGET995', 'TARGET996', 'TARGET997', 'TARGET998', 'TARGET999']
`Time taken 2.82 seconds to delete 1000 items`

Process finished with exit code 0

**Reply by harshitsinghai77** [reply]
hmm.. not sure why I'm facing so much time with the same code and less number of entries. Anyway, thanks for the help. For me, it still takes lots of time.

can you clarify how your `ep = self.__base_root + base_name + '/items/' + key` and `self.__session` looks like?
You're not passing the session?

I'm doing something like
```python
async def delete_task_http(session, key):
        url = (
            f"https://database.deta.sh/v1/{PROJECT_ID}/{DETA_TASK_BASENAME}/items/{key}"
        )
        await session.delete(url, headers=headers)
        print("Done", url)
        return url

async def delete_all_tasks(cls, user_key: str):
        """Get all the tasks and then make multiple delete request concurrently using asyncio."""
        # at the time of writing, deta doesn't support deleteMany.
        all_tasks = cls.get_tasks(user_key)
        all_tasks = [task["key"] for task in all_tasks]

        async with aiohttp.ClientSession() as session:
            task = [
                asyncio.create_task(cls.delete_task_http(session, key))
                for key in all_tasks
            ]
            await asyncio.gather(*task)
            print("All task deleted")
        
```

**Reply by jnsougata** [reply]
I think `all_tasks = cls.get_tasks(user_key)` takes the most of the time! can you show how it is defined or did you test how much time it takes?
